File: src/09.go
package main

import (
	"fmt"
	"sort"

	. "aoc2021/src/common"
)

func solve_09_1(floorMap floorMap) int {
	result := 0
	for _, point := range floorMap.findLowPoints() {
		result += floorMap.riskAt(point)
	}
	return result
}

func solve_09_2(floorMap floorMap) any {
	var basinSizes []int
	for _, point := range floorMap.findLowPoints() {
		basinSizes = append(basinSizes, floorMap.basinSizeAt(point))
	}
	sort.Sort(sort.Reverse(sort.IntSlice(basinSizes)))
	return basinSizes[0] * basinSizes[1] * basinSizes[2]
}

type floorMap struct {
	lines []string
	w, h  int
}

func (self *floorMap) findLowPoints() []floorPoint {
	var result []floorPoint
	for i := 0; i < self.h; i++ {
		for j := 0; j < self.w; j++ {
			point := floorPoint{i, j}
			if self.isLowPoint(point) {
				result = append(result, point)
			}
		}
	}
	return result
}

func (self *floorMap) isLowPoint(point floorPoint) bool {
	c := self.charAt(point)
	for _, n := range self.neighbors(point) {
		if self.charAt(n) <= c {
			return false
		}
	}
	return true
}

func (self *floorMap) riskAt(point floorPoint) int {
	return ParseDigit(self.charAt(point)) + 1
}

func (self *floorMap) basinSizeAt(lowPoint floorPoint) int {
	result := 0
	known := map[floorPoint]bool{}
	var q []floorPoint

	process := func(point floorPoint) {
		result += 1
		Enqueue(&q, point)
		known[point] = true
	}

	process(lowPoint)
	for len(q) > 0 {
		point := Dequeue(&q)
		for _, n := range self.flowingNeighbors(point) {
			if !known[n] {
				process(n)
			}
		}
	}
	return result
}

func (self *floorMap) flowingNeighbors(point floorPoint) []floorPoint {
	c := self.charAt(point)

	var result []floorPoint
	for _, n := range self.neighbors(point) {
		neighborChar := self.charAt(n)
		if c < neighborChar && '9' != neighborChar {
			result = append(result, n)
		}
	}
	return result
}

func (self *floorMap) neighbors(point floorPoint) []floorPoint {
	var result []floorPoint
	if 0 < point.i {
		result = append(result, floorPoint{point.i - 1, point.j})
	}
	if point.i < self.h-1 {
		result = append(result, floorPoint{point.i + 1, point.j})
	}
	if 0 < point.j {
		result = append(result, floorPoint{point.i, point.j - 1})
	}
	if point.j < self.w-1 {
		result = append(result, floorPoint{point.i, point.j + 1})
	}
	return result
}

func (self *floorMap) charAt(point floorPoint) byte {
	return self.lines[point.i][point.j]
}

type floorPoint struct {
	i, j int
}

func main() {
	lines := GetInputLines()
	floorMap := floorMap{lines, len(lines[0]), len(lines)}
	fmt.Println(solve_09_1(floorMap))
	fmt.Println(solve_09_2(floorMap))
}

File: src/09.js
// @ts-check
import { readFileSync } from 'fs'

function solve_09_1(floorMap) {
  let result = 0
  for (const point of floorMap.findLowPoints()) {
    result += floorMap.riskAt(point.i, point.j)
  }
  return result
}

class FloorMap {
  constructor(lines) {
    this.lines = lines
    this.w = lines[0].length
    this.h = lines.length
  }

  findLowPoints() {
    const result = []
    for (let i = 0; i < this.h; i++) {
      for (let j = 0; j < this.w; j++) {
        if (this.isLowPoint(i, j)) {
          result.push({i: i, j: j})
        }
      }
    }
    return result
  }

  isLowPoint(i, j) {
    const c = this.charAt(i, j)
    for (const n of this.neighbors(i, j)) {
      if (this.charAt(n.i, n.j) <= c) {
        return false
      }
    }
    return true
  }

  charAt(i, j) {
    return this.lines[i][j]
  }

  neighbors(i, j) {
    const result = []
    if (0 < i) {
      result.push({i: i-1, j})
    }
    if (i < this.h-1) {
      result.push({i: i+1, j})
    }
    if (0 < j) {
      result.push({i, j: j-1})
    }
    if (j < this.w-1) {
      result.push({i, j: j+1})
    }
    return result
  }

  riskAt(i, j) {
    return 1+ Number(this.charAt(i, j))
  }
}

const file = readFileSync(process.argv[2], 'utf8')
const lines = file.split('\n').slice(0, -1)
const floorMap = new FloorMap(lines)

console.log(solve_09_1(floorMap))
